components/places/src/api/places_api.rs

```diff
@@ -30,6 +30,34 @@ use sync15::{sync_multiple, telemetry, MemoryCachedState, SyncResult};
 // per collection.
 pub const GLOBAL_STATE_META_KEY: &str = "global_sync_state_v2";
 
+// Our "sync manager" will use whatever is stashed here.
+lazy_static::lazy_static! {
+    // Mutex: just taken long enough to update the contents - needed to wrap
+    //        the Weak as it isn't `Sync`
+    // [Arc/Weak]: Stores the places api used to create the connection for
+    //             BookmarksEngine/HistoryEngine
+    static ref PLACES_API_FOR_SYNC_MANAGER: Mutex<Weak<PlacesApi>> = Mutex::new(Weak::new());
+}
+
+// Called by the sync manager to get a sync engine via the PlacesApi previously
+// registered with the sync manager.
+pub fn get_registered_sync_engine(name: &str) -> Option<Box<dyn SyncEngine>> {
+    match PLACES_API_FOR_SYNC_MANAGER.lock().unwrap().upgrade() {
+        None => None,
+        Some(places_api) => match places_api.get_sync_connection() {
+            Ok(db) => match name {
+                "bookmarks" => Some(Box::new(BookmarksSyncEngine::new(db))),
```

Comment:
The reason we don't need to `Arc::clone` here is because `Weak::upgrade` returns a **new** strong reference right? This makes sense, so every time we call `get_sync_connection` and it returns the same connection, it'll add a reference count so we're good!

(I'm just speaking out loud because I had to go look at docs once I saw that we don't clone the Arc here)

Reply:
Yes I'm pretty sure that's how it works.  I used to have a clone there, but I deleted it and it still compiled so I think we're good :)